File: 2024/d04/src/main.py
input = open("src/input").read()

test = """MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX"""  # 18


def part1(grid: list[str]):
    word = "XMAS"
    adjacent = [(-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1)]
    out = 0
    for r in range(len(grid)):
        for c in range(len(grid[0])):
            if grid[r][c] == word[0]:
                for ar, ac in adjacent:
                    if not (
                        0 <= r + ar * len(word[1:]) < len(grid)
                        and 0 <= c + ac * len(word[1:]) < len(grid[0])
                    ):
                        continue

                    if all(
                        grid[r + ar * i][c + ac * i] == word[i]
                        for i in range(1, len(word))
                    ):
                        out += 1
    return out


def part2(grid: list[str]):
    word = "MAS"
    out = 0
    for r in range(1, len(grid) - 1):
        for c in range(1, len(grid[0]) - 1):
            if grid[r][c] != word[1]:
                continue

            if "".join(
                grid[r + ar][c + ac]
                for ar, ac in [(-1, -1), (-1, 1), (1, 1), (1, -1)]
                # M . S
                # . A .
                # M . S
            ) in ["MSSM", "SSMM", "SMMS", "MMSS"]:
                out += 1
    return out


grid = input.splitlines()
print(f"{part1(grid)=}")
print(f"{part2(grid)=}")

File: 2024/d04/src/main.rs
const INPUT: &str = include_str!("input");

const _TEST: &str = "MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX";

fn part1(grid: &[Vec<char>]) -> i32 {
    let word = ['X', 'M', 'A', 'S'];
    let mut out = 0;
    for r in 0..grid.len() {
        for c in 0..grid[0].len() {
            if grid[r][c] != word[0] {
                continue;
            }
            for ar in [-1, 0, 1] {
                for ac in [-1, 0, 1] {
                    if ar == 0 && ac == 0 {
                        continue;
                    }
                    let rest = word[1..].len() as isize;
                    let maxr = r as isize + ar * rest;
                    let maxc = c as isize + ac * rest;
                    if !((0 <= maxr && maxr < grid.len() as isize)
                        && (0 <= maxc && maxc < grid[0].len() as isize))
                    {
                        continue;
                    }
                    if (1..word.len()).all(|i| {
                        grid[(r as isize + ar * i as isize) as usize]
                            [(c as isize + ac * i as isize) as usize]
                            == word[i]
                    }) {
                        out += 1;
                    }
                }
            }
        }
    }
    out
}

fn part2(grid: &[Vec<char>]) -> i32 {
    let word = ['M', 'A', 'S'];
    let mut out = 0;
    for r in 1..grid.len() - 1 {
        for c in 1..grid[0].len() - 1 {
            if grid[r][c] != word[1] {
                continue;
            }

            let edges = [(-1, -1), (-1, 1), (1, 1), (1, -1)]
                .map(|(ar, ac)| grid[(r as isize + ar) as usize][(c as isize + ac) as usize]);

            // M . S
            // . A .
            // M . S
            if [
                ['M', 'S', 'S', 'M'],
                ['S', 'S', 'M', 'M'],
                ['S', 'M', 'M', 'S'],
                ['M', 'M', 'S', 'S'],
            ]
            .iter()
            .any(|&pat| pat == edges)
            {
                out += 1;
            }
        }
    }
    out
}

fn main() {
    let grid: Vec<Vec<char>> = INPUT.lines().map(|line| line.chars().collect()).collect();
    println!("part1 = {}", part1(&grid));
    println!("part2 = {}", part2(&grid));
}
